File: it-agents/skills/code-architect/references/code-examples/csharp/Proxy.cs
// Proxy — a placeholder controlling access to a real object (here: lazy creation + caching).
using System;
using System.Collections.Generic;

namespace Patterns.Proxy
{
    interface IService
    {
        string Fetch(string key);
    }

    // Expensive-to-build real subject.
    sealed class RealService : IService
    {
        public RealService() => Console.WriteLine("[RealService constructed]");
        public string Fetch(string key) => $"value-for-{key}";
    }

    // Proxy defers construction and caches results.
    sealed class CachingProxy : IService
    {
        private RealService? _real;
        private readonly Dictionary<string, string> _cache = new();

        public string Fetch(string key)
        {
            if (_cache.TryGetValue(key, out string? cached))
                return $"{cached} (cached)";

            _real ??= new RealService(); // lazy init on first real need
            string result = _real.Fetch(key);
            _cache[key] = result;
            return result;
        }
    }

    static class Demo
    {
        public static void Run()
        {
            IService service = new CachingProxy();
            Console.WriteLine("Proxy created; no real service yet.");
            Console.WriteLine(service.Fetch("x"));
            Console.WriteLine(service.Fetch("x"));
            Console.WriteLine(service.Fetch("y"));
        }

        public static void Main() => Run();
    }
}

File: it-agents/skills/code-architect/references/code-examples/swift/Proxy.swift
// Proxy — a placeholder controlling access to a real object (here: lazy + caching).

protocol Resource {
    func load() -> String
}

// Expensive real subject.
final class HeavyResource: Resource {
    private let id: String
    init(id: String) {
        self.id = id
        print("HeavyResource(\(id)) constructed")
    }
    func load() -> String { "payload for \(id)" }
}

// Proxy defers construction and caches the result.
final class ResourceProxy: Resource {
    private let id: String
    private var real: HeavyResource?
    private var cached: String?

    init(id: String) { self.id = id }

    func load() -> String {
        if let cached { return cached + " (cached)" }
        if real == nil { real = HeavyResource(id: id) }
        let result = real!.load()
        cached = result
        return result
    }
}

func runDemo() {
    let proxy = ResourceProxy(id: "doc-1")
    print("proxy created; nothing loaded yet")
    print(proxy.load())
    print(proxy.load())
}

runDemo()
